avldb/src/main/scala/scorex/crypto/authds/avltree/batch/VersionedLDBAVLStorage.scala

```diff
@@ -64,83 +69,175 @@ class VersionedLDBAVLStorage[D <: Digest](store: LDBVersionedStore,
-          value -> nextLeafKey
-        }
-        val l = new ProverLeaf[D](key, value, nextLeafKey)
-        l.isNew = false
-        l
+
+  private[batch] val topNodeHashKey: Array[Byte] = Array.fill(StateTreeParameters.labelSize)(123: Byte)
+  private[batch] val topNodeHeightKey: Array[Byte] = Array.fill(StateTreeParameters.labelSize)(124: Byte)
+
+  // prefixes used to encode node type (internal or leaf) in database
+  private[batch] val InternalNodePrefix: Byte = 0: Byte
+  private[batch] val LeafPrefix: Byte = 1: Byte
+
+  /**
+    * Node serializer which is restoring nodes with only children labels stored in, without possibility to restore
+    * children automatically when requested
+    */
+  val noStoreSerializer = new ProverNodeSerializer(store = null)
+
+  /**
+    * Fetch tree node from database by its database id (hash of node contents)
+    *
+    * @param dbKey - database key node of interest is stored under (hash of the node)
+    * @param store - database
+    * @return node read from the database (or throws exception if there is no such node), in case of internal node it
+    *         returns pruned internal node, so a node where left and right children not stored, only their hashes
+    */
+  def fetch(dbKey: ADKey)(store: LDBVersionedStore): ProverNodes[DigestType] = {
+    val bytes = store(dbKey)
+    val node = new ProverNodeSerializer(store).parseBytes(bytes)
+    node.isNew = false
+    node
+  }
+
+  /**
+    * Node label (hash). Used as database key for node contents
+    */
+  private[batch] def nodeLabel(node: ProverNodes[DigestType]): Array[Byte] = node.label
+
+  /**
+    * Reconstruct versioned AVL+ tree storage from dump
+    * @param manifest - tree started from root node cut at fixed depth
+    * @param chunks - iterator (lazy sequence) of subtrees under manifest tree
+    * @param additionalData - additional fields (metadata) to write into database
+    * @param store - db to store restored tree to
+    * @return db with reconstructed tree written to, or error
+    */
+  def recreate(manifest: BatchAVLProverManifest[DigestType],
+               chunks: Iterator[BatchAVLProverSubtree[DigestType]],
+               additionalData: Iterator[(Array[Byte], Array[Byte])],
+               store: LDBVersionedStore): Try[VersionedLDBAVLStorage] = {
+    //todo: the function below copy-pasted from BatchAVLProver, eliminate boilerplate?
+
+    def idCollector(node: ProverNodes[DigestType],
+                    acc: Iterator[(Array[Byte], Array[Byte])]): Iterator[(Array[Byte], Array[Byte])] = {
+      val pair: (Array[Byte], Array[Byte]) = (nodeLabel(node), noStoreSerializer.toBytes(node))
+      node match {
+        case n: ProxyInternalNode[DigestType] if n.isEmpty =>
+          acc ++ Iterator(pair)
+        case i: InternalProverNode[DigestType] =>
+          acc ++ Iterator(pair) ++ idCollector(i.left, acc) ++ idCollector(i.right, acc)
```

Comment:
Please provide an alternative, otherwise, there is no constructive talk here.

Reply:
Use TraversableOnce for example. The recursive method is suspicious, have you tested in with unit tests?
What do you think this code prints?
    val it = Iterator[Int](1)
    val res = it ++ Iterator(2) ++ it
    println(res.toList)

Reply:
I think the code will work iff the acc is empty. In this case the acc parameter is misleading and can be removed.

Reply:
My code prints List(1, 2) instead List(1, 2, 1) if Iterator were immutable.
In you code acc is always empty iterator, which is why it works.
This may be the fix https://github.com/ergoplatform/ergo/pull/1977